### ISSUE TITLE
Support Server-Sent Events (SSE) with adapter-node

### ISSUE THREAD

**Issue by babeard** [sveltejs/kit#887]
**Describe the bug**
While migrating a sapper app to svelte-kit / adapter-node, I came across a feature that seems to be missing. Namely, the ability to create custom middleware that can handle [Server-Sent Events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events) or other long-lived connections that send data to the response stream without closing the connection. Since the handle hook requires **returning** `Response | Promise<Response>`, thus closing the connection, it does not support this. Nor have I found any other way to `res.write();` using endpoints.

**Information about your SvelteKit Installation:**

<details>
  <summary>Diagnostics</summary>

```
System:
    OS: Linux 4.4 Ubuntu 20.04.1 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz
    Memory: 7.97 GB / 15.94 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
Binaries:
    Node: 14.5.0 - ~/.nvm/versions/node/v14.5.0/bin/node
    npm: 6.14.5 - ~/.nvm/versions/node/v14.5.0/bin/npm
npmPackages:
    @sveltejs/kit: ^1.0.0-next.68 => 1.0.0-next.68
    svelte: ^3.37.0 => 3.37.0
    vite: ^2.1.5 => 2.1.5
```
- Using node-adapter
</details>

**Severity**
Adds a bit of friction for migrating from Sapper, reduces feature parity, and will require me to either 
a) regress to polling the server from the client, or 
b) write a custom adapter for every project that needs SSEs. (which I'm not sure that this even helps for the development environment anyways)

**Additional context**
[Issue #334](https://github.com/sveltejs/kit/issues/334) seems relevant. 

Also originally asked within [discord chat](https://discord.com/channels/457912077277855764/819723698415599626/827550800003465216) but no one seems to know the answer, which is why I am submitting this issue.

**Comment by cayter** [comment]
@babeard After going through https://vitejs.dev/guide/api-plugin.html#configureserver, I believe we can try the below:

1. Come up with a middleware with the method signature `function (req, res, next) => {}`.
2. Configure the middleware in `svelte.config.cjs`'s `vite.plugins`. (this is for `svelte-kit dev`)
3. Come up with a forked `adapter-node` that allows polka to use the same middleware. (this is for `node build/index.js`)

I haven't tested it but I think it should work.

**Comment by mromanuk** [comment]
> @babeard After going through https://vitejs.dev/guide/api-plugin.html#configureserver, I 

I managed to make it work with curl, but I have no luck trying to call `/messages` from a regular svelte component using `EventSource` doing something like this

```svelte
<script>
	import { onMount } from 'svelte'
	
	onMount(async () => {
		const sse = new EventSource('/messages')

		sse.onmessage = function (e) {
			console.log(e.data)
		}
	})

</script>
```

Would you mind pointing me in the right direction?

**Comment by mromanuk** [comment]
@cayter Do you have any  recommendation on how to patch `build/index.js` to append the SSE middleware?

**Comment by cayter** [comment]
> @cayter Do you have any recommendation on how to patch `build/index.js` to append the SSE middleware?

I checked the [src](https://github.com/sveltejs/kit/blob/master/packages/adapter-node/index.js), it isn't possible at all which means we will have to change how `adapter-node` works. For now, you can only write a `postbuild` npm script to find and replace the pattern in `build/index.js`.

**Comment by mromanuk** [comment]
yes, thank you! It's not pretty, but it works.

**Comment by mromanuk** [comment]
I ended up adding this to `package.json`, `sse` is the function exported by vite, because it's imported in `hook.js`

```
    "postbuild": "sed -i 's/polka().use(/polka().use(sse, /g' build/index.js",
```

**Comment by cayter** [comment]
@mromanuk 

I had come up with a better solution and it's gonna be a monolith like Rails, the toolkit source code is [here](https://github.com/appist/kit). Note that this isn't production ready but I plan to continue adding more capabilities to make it easier for ppl to build and deploy app easily just like how Rails/Laravel is.

Project Demo: https://codesandbox.io/s/appistkit-demo-4w58d
Project Folder Structure:
```sh
.
├── README.md
├── app (this is a shell script where we will use to run all the app commands, you can run `./app --help` to see what commands are available)
├── configs (12factor app, running `KIT_ENV=staging ./app server` would lead the app to load `configs/.env.staging`)
├── package-lock.json
├── package.json
├── src
│   ├── client (this is the SvelteKit folder)
│   │   ├── app.html
│   │   ├── global.d.ts
│   │   ├── pages
│   │   │   ├── __error.svelte
│   │   │   ├── __layout.svelte
│   │   │   └── index.svelte
│   │   └── static
│   │       └── favicon.png
│   ├── cmd
│   │   └── now.ts
│   ├── db (this is the DB migrate/seed folder like what Rails have)
│   │   ├── migrate
│   │   │   └── primary
│   │   │       └── 20210422155916_create_users.ts
│   │   └── seed
│   │       └── primary
│   │           └── index.ts
│   ├── server (this is the server folder that is also file-based routing like SvelteKit but for server-side, you can write your SSE handler in any path you want, I'm planning to add `export const ws` to support websocket and `export const graphql` to support GraphQL and probably `export const grpc` to support GRPC too)
│   │   └── api.ts
│   └── worker (this is the worker folder where we would store the BullMQ worker logic, WIP)
├── svelte.config.js
└── tsconfig.json
```

Let me know what you think. Thanks.

**Comment by cayter** [comment]
@mromanuk 

I had come up with a better solution and it's gonna be a monolith like Rails, the toolkit source code is [here](https://github.com/appist/kit). Note that this isn't production ready but I plan to continue adding more capabilities to make it easier for ppl to build and deploy app easily just like how Rails/Laravel is.

Project Demo: https://codesandbox.io/s/appistkit-demo-4w58d
Project Folder Structure:
```sh
.
├── README.md
├── app (this is a shell script where we will use to run all the app commands, you can run `./app --help` to see what commands are available)
├── configs (12factor app, running `KIT_ENV=staging ./app server` would lead the app to load `configs/.env.staging`, I plan to add the encrypt/decrypt capability with KIT_MASTER_KEY and separate the server/client as client is restricted to use `VITE_` prefix)
├── package-lock.json
├── package.json
├── src
│   ├── client (this is the SvelteKit folder)
│   │   ├── app.html
│   │   ├── global.d.ts
│   │   ├── pages
│   │   │   ├── __error.svelte
│   │   │   ├── __layout.svelte
│   │   │   └── index.svelte
│   │   └── static
│   │       └── favicon.png
│   ├── cmd (this is the command folder which allows you to easily add new commands to `./app`)
│   │   └── now.ts
│   ├── db (this is the DB migrate/seed folder like what Rails have)
│   │   ├── migrate
│   │   │   └── primary
│   │   │       └── 20210422155916_create_users.ts
│   │   └── seed
│   │       └── primary
│   │           └── index.ts
│   ├── server (this is the server folder that is also file-based routing like SvelteKit but for server-side, you can write your SSE handler in any path you want, I'm planning to add `export const ws` to support websocket and `export const graphql` to support GraphQL and probably `export const grpc` to support GRPC too)
│   │   └── api.ts
│   └── worker (this is the worker folder where we would store the BullMQ worker logic, WIP)
├── svelte.config.js
└── tsconfig.json
```

The Polka instance used in: 

- `./app dev` (for local development)
- `NODE_ENV=production KIT_ENV=staging ./app server` (for running on staging/production environments, this command is only usable after you run `./app build` to generate the `dist` folder) 

are exactly the same instance. Let me know what you think. Thanks.

**Comment by cayter** [comment]
@mromanuk 

I had come up with a better solution and it's gonna be a monolith like Rails, the toolkit source code is [here](https://github.com/appist/kit). Note that this isn't production ready but I plan to continue adding more capabilities to make it easier for ppl to build and deploy app easily just like how Rails/Laravel is.

Project Demo: https://codesandbox.io/s/appistkit-demo-4w58d
Project Folder Structure:
```sh
.
├── README.md
├── app (this is a shell script where we will use to run all the app commands, you can run `./app --help` to see what commands are available)
├── configs (12factor app, running `KIT_ENV=staging ./app server` would lead the app to load `configs/.env.staging`, I plan to add the encrypt/decrypt capability with KIT_MASTER_KEY and separate the server/client as client is restricted to use `VITE_` prefix)
├── package-lock.json
├── package.json
├── src
│   ├── client (this is the SvelteKit folder)
│   │   ├── app.html
│   │   ├── global.d.ts
│   │   ├── pages
│   │   │   ├── __error.svelte
│   │   │   ├── __layout.svelte
│   │   │   └── index.svelte
│   │   └── static
│   │       └── favicon.png
│   ├── cmd (this is the command folder which allows you to easily add new commands to `./app`)
│   │   └── now.ts
│   ├── db (this is the DB migrate/seed folder like what Rails have)
│   │   ├── migrate
│   │   │   └── primary
│   │   │       └── 20210422155916_create_users.ts
│   │   └── seed
│   │       └── primary
│   │           └── index.ts
│   ├── server (this is the server folder that is also file-based routing like SvelteKit but for server-side, you can write your SSE handler in any path you want, I'm planning to add `export const ws` to support websocket and `export const graphql` to support GraphQL and probably `export const grpc` to support GRPC too)
│   │   └── api.ts
│   └── worker (this is the worker folder where we would store the BullMQ worker logic, WIP)
├── svelte.config.js
└── tsconfig.json
```

The Polka instance used in: 

- `./app dev` (for local development)
- `NODE_ENV=production KIT_ENV=staging ./app server` (for running on staging/production environments, this command is only usable after you run `./app build` to generate the `dist` folder) 

are exactly the same instance.

Originally, I thought of including the server logic as part of the client (SvelteKit) folder but had bumped into lots of bundling issues that Vite has yet to resolve. So to keep things separated in cleaner way, I decided to just have a `client` and `server` folder. The current solution is like seamlessly combining your traditional NodeJS server including the best out of what SvelteKit is providing (i.e. great frontend DX, SSR, SSG and etc).

Let me know what you think. Thanks.

**Comment by jsprog** [comment]
Thanks to both @cayter and @babeard  for the information provided above.

For development mode, I already followed the instructions from @cayter, but for builds, I had to avoid external scripts (postbuild) and stick to javascript only.

`src/hooks.js`
```js
// you many not use 'sse_middleware' inside src/hooks.js, but
// - it has to land in the final bundle during builds.
// - we need to ensure automated server's reload during development
// - the re-export is necessary to avoid dead code removal during builds
export { default as sse_middleware } from '$lib/rtc/sse_middleware'

```

`svelte.config.js`
```js
import adapter_node from '@sveltejs/adapter-node'
import fs from 'fs'
import path, { resolve } from 'path'
import preprocess from 'svelte-preprocess';
import sse_middleware from './src/lib/rtc/sse_middleware.js'

/** @type {import('@sveltejs/kit').Config} */
const config = {
    preprocess: preprocess(),
    
    kit: {
        // hydrate the <div id="svelte"> element in src/app.html
        target: '#svelte',

        adapter: {
            async adapt (opts) {
                const adapter = adapter_node()
                await adapter.adapt(opts)

                // build/index.js: attach the middleware to polka instance (text replacement)
                let content = fs.readFileSync(path.resolve('./build/index.js'), 'utf-8')
                content = content.replace(`polka().use(`, `polka().use(sse_middleware).use(`)
                fs.writeFileSync('./build/index.js', content, 'utf-8')
            }
        }, 

        vite:  {
            // temporary fix for using broadcast-channel
            define: { 'process.browser': true },

            plugins: [
                { name: 'sse_middleware', configureServer (server) {
                    server.middlewares.use(sse_middleware)
                }}
            ],

            build: {
                rollupOptions: {
                    plugins: [
                        {
                            name: 'postbuild-fixes',
                            writeBundle () {
                                // at first attempt, I tried to attach the middleware to polka instance here,
                                // unfortunately, build/index.js isn't available at this stage
                            }
                        }
                    ]
                }
            }

        },

    }
};

export default config;

```

**Comment by cayter** [comment]
> Thanks to both @cayter and @mromanuk for the information provided above.
> 
> For development mode, I already followed the instructions from @cayter, but for builds, I had to avoid external scripts (postbuild) and stick to javascript only.
> 
> `src/hooks.js`
> 
> ```js
> // you many not use 'sse_middleware' inside src/hooks.js, but
> // - it has to land in the final bundle during builds.
> // - we need to ensure automated server's reload during development
> // - the re-export is necessary to avoid dead code removal during builds
> export { default as sse_middleware } from '$lib/rtc/sse_middleware'
> ```
> 
> `svelte.config.js`
> 
> ```js
> import adapter_node from '@sveltejs/adapter-node'
> import fs from 'fs'
> import path, { resolve } from 'path'
> import preprocess from 'svelte-preprocess';
> import sse_middleware from './src/lib/rtc/sse_middleware.js'
> 
> /** @type {import('@sveltejs/kit').Config} */
> const config = {
>     preprocess: preprocess(),
>     
>     kit: {
>         // hydrate the <div id="svelte"> element in src/app.html
>         target: '#svelte',
> 
>         adapter: {
>             async adapt (opts) {
>                 const adapter = adapter_node()
>                 await adapter.adapt(opts)
> 
>                 // build/index.js: attach the middleware to polka instance (text replacement)
>                 let content = fs.readFileSync(path.resolve('./build/index.js'), 'utf-8')
>                 content = content.replace(`polka().use(`, `polka().use(sse_middleware).use(`)
>                 fs.writeFileSync('./build/index.js', content, 'utf-8')
>             }
>         }, 
> 
>         vite:  {
>             // temporary fix for using broadcast-channel
>             define: { 'process.browser': true },
> 
>             plugins: [
>                 { name: 'attach_middlewares', configureServer (server) {
>                     server.middlewares.use(sse_middleware)
>                 }}
>             ],
> 
>             build: {
>                 rollupOptions: {
>                     plugins: [
>                         {
>                             name: 'postbuild-fixes',
>                             writeBundle () {
>                                 // at first attempt, I tried to attach the middleware to polka instance here,
>                                 // unfortunately, build/index.js isn't available at this stage
>                             }
>                         }
>                     ]
>                 }
>             }
> 
>         },
> 
>     }
> };
> 
> export default config;
> ```

@jsprog The `postbuild` script mentioned here is the `scripts.postbuild` in `package.json` which will run after `scripts.build`.

**Comment by jsprog** [comment]
The modules imported by `svelte.config.js`, are not part of the flow receiving all kinds of transformations by sveltekit. You'll miss **aliases**, **HMR**, and even **access to imports without providing the extensions (even for index.js)**.

Example of a file imported by `svelte.config.js`
```js
import db from '$lib/database'     // fails
import db from './lib/database'    // fails
import db from './lib/database.js' // works
```

Instead, I had to resort to **global**

`src/lib/rtc/server/middlewares/rtc_sse`
```js
// global.rtc_sse will be used by configuration for vite (svelte.config.js)
// this also ensure that no dead code is removed for builds
global.rtc_sse = rtc_sse

export default async function rtc_sse (req, res, next) {
    // ...
}
```

`src/hooks.js`
```js
import '$lib/rtc/server/middlewares/rtc_sse'

```

`svelte.config.js`
```js

  vite: {
    plugins: [
      { name: 'use-middlewares', configureServer (server) {
        server.middlewares.use((req, res, next) => {
          global.rtc_sse ? global.rtc_sse(req, res, next) : next()
        })
      }}
    ]
  }
```

**Comment by christophemacabiau** [comment]
Struggling also with this issue :-)
Thanks to @cayter @mromanuk & @jsprog  for their suggestions

To get it working in prod, I need to import bodyParser, otherwise we don't have access to the posted data. My solution was to add
```ts
global.bodyParser = bodyParser;
```
in the middleware and to update the postbuild script (https://github.com/sveltejs/kit/issues/887#issuecomment-861718102).

I wonder if any of you has a better solution?

**Comment by mustofa-id** [comment]
as this #3384 changes, are now possible to implement SSE through endpoint?

**Comment by mustofa-id** [comment]
> as this #3384 changes, are now possible to implement SSE through endpoint?

I found it works, Idk if this correct.

```js
import { PassThrough } from 'stream'

export const get: RequestHandler = async () => {
	const channel = new PassThrough

	const timer = setInterval(()=> {
		channel.write(`event: timer\ndata: ${new Date().toLocaleString()}\n\n`)
	}, 1000)

	channel.on('close', () => { // not get called on client close
		clearInterval(timer)
	})

	return {
		headers: {
			'Cache-Control': 'no-cache',
			'Content-Type': 'text/event-stream',
			'Connection': 'keep-alive'
		},
		body: channel
	}
}
```

**Comment by mromanuk** [comment]
> as this #3384 changes, are now possible to implement SSE through endpoint?

Is that now possible?

**Comment by Rich-Harris** [comment]
It's not possible _today_, but it's definitely something we want to get fixed for 1.0. This is basically just a special case of #3419. It _should_ be possible to do something like this, I think...

```js
const controllers = new Set();

export function get() {
  let controller;

  return {
    body: new ReadableStream({
      start: (_) => {
        controller = _;
        controllers.add(controller);
      },
      cancel: () => {
        controllers.delete(controller);
      }
    })
  };
}

export async function post({ request }) {
  const message = await request.text();

  // importantly, this use case wouldn't work in serverless environments —
  // for these sorts of use cases we need things like this:
  //   https://deno.com/deploy/docs/runtime-broadcast-channel
  //
  // ideally Kit would have some kind of abstraction that leveraged
  // BroadcastChannel or WebSockets or whatever under the hood
  for (const controller of controllers) {
    controller.enqueue(message);
  }

  return { status: 204 };
}
```

...but `node-fetch` breaks spec by using [`stream.Readable`](https://nodejs.org/api/stream.html#readable-streams) instead of [`ReadableStream`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream). Given that Node itself is gradually embracing `fetch` and web streams, supporting `stream.Readable` is an evolutionary dead end (it almost works today in SvelteKit, but it's buggy — doesn't support request cancellation — and only works locally and in Node-like production environments, whereas we want these things to work cross-platform), so I suspect the answer is to [polyfill](https://www.npmjs.com/package/web-streams-polyfill) web streams in Node.

I'll close this in favour of #3419.
